Change Superbuild file hierarchy

TheodoreDavis opened the pull request:
Have CMakePublic build inside of ../helloworld/external
Allow find_package to find CMakePublic's config

Commit:
- Removed demo/ from .gitignore
- ./CMakeList
-- Removed extra slash from stage_directory
-- Grab from TheodoreDavis/CMakePublic instead for now
-- Change install paths on CMakePublic to move it inside the helloworld dir
-- Add HelloWorld dir to CMAKE_PREFIX_PATH
- hello_world/CMakeList
-- Remove debug message

TheodoreDavis commented:
Tests were able to find the CMakePublic package. Had to delete the ones on my local system since it looks there as well. Right now it finds the package with line #32 in ./CMakeLists.txt

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -2,27 +2,35 @@ cmake_minimum_required(VERSION 3.5.1)
     CMakePublic
-    GIT_REPOSITORY "https://github.com/CMaizeExamples/CMakePublic"
+    #GIT_REPOSITORY "https://github.com/CMaizeExamples/CMakePublic"
+    GIT_REPOSITORY "https://github.com/TheodoreDavis/CMakePublic"
```
Is this just for testing? Or are there changes needed in CMakePublic?

ryanmrichard commented:
At least on my system there seems to be some problem with how CMAKE_PREFIX_PATH gets passed into HelloWorld. More specifically it looks like only the first path is getting passed in to it (also the first `list(append CMAKE_PREFIX_PATH...)` call seems to assume that there is at least one element in the list already). One solution I am aware of is to use CMAKE_CACHE_ARGS (instead of CMAKE_ARGS) in the `ExternalProject_Add` command. This doesn't butcher the list. It may be possible to pass a list to CMAKE_ARGS, but it may require escapes.

Commit:
Fix error with lists not getting passed to external project

Commit:
Make external path into a variable for ease of use

Commit:
- Fix include path of cmakepublic to new dir
- Add a return at the end of SUBDIRS command for testing

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -2,37 +2,50 @@ cmake_minimum_required(VERSION 3.5.1)
 ExternalProject_Add(
     "${PROJECT_NAME}_core"
     SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/hello_world"
+    LIST_SEPARATOR |
```
This is neat. Didn't know about this.

ryanmrichard commented on hello_world/src/hello-world.cpp:
```diff
@@ -1,5 +1,5 @@
 #include "hello-world.hpp"
-#include "cmakepublic/cmake_public.hpp"
+#include "helloworld/external/cmakepublic/cmake_public.hpp"
```
```suggestion
#include "cmakepublic/cmake_public.hpp"

```
We don't want to put the `helloworld/external` prefix on there because if the user provides their own copy of CMakePublic, it's not going to be in that directory. When we look for CMakePublic inside the hello world external project, the resulting `cmakepublic` target should have its include paths set up so that `#include "cmakepublic/cmake_public.hpp"`

Commit:
Fixed header inclusion path errors
- Added extra tar_inc_dir for CMakePublic
- Fixed issue with include paths in hello-world.cpp
- Fixed issue with include paths in test.cpp

ryanmrichard commented on hello_world/CMakeLists.txt:
```diff
@@ -28,7 +28,14 @@ set_target_properties(${PROJECT_NAME}
+target_include_directories(CMakePublic::CMakePublic
+    INTERFACE
+        $<BUILD_INTERFACE:${CMAKE_INSTALL_FULL_INCLUDEDIR}/helloworld>
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/helloworld>
```
you shouldn't have to set these. They should be on the target you get from `find_package`

ryanmrichard commented on hello_world/src/hello-world.cpp:
```diff
@@ -1,5 +1,5 @@
-#include "hello-world.hpp"
-#include "helloworld/external/cmakepublic/cmake_public.hpp"
+#include <hello-world.hpp>
+#include <external/cmakepublic/cmake_public.hpp>
```
```suggestion
#include "helloworld/hello-world.hpp"
#include <cmakepublic/cmake_public.hpp>
```
(usual C/C++ convention is to use double quotes for header files which are part of the current project and angle brackets for header files from other projects). Getting the first line to work will also require adding a namespace directory `helloworld` under the `include` directory. Not sure why it wasn't there to begin with.

As for the remainder of the suggestion:

IIRC `CMakePublic` installs the header file to `CMAKE_INSTALL_PREFIX/include/cmakepublic/cmake_public.hpp` and when you do `find_package(cmakepublic)` the resulting target's include path will be the `CMAKE_INSTALL_PREFIX/include` part (here `CMAKE_INSTALL_PREFIX` is what was passed to `CMakePublic`, and will be different than the `CMAKE_INSTALL_PREFIX` being passed to HelloWorld; also I can't remember off the top of my head whether the include path has the `cmakepublic` directory on it or not). 

I forget exactly what path we were giving `CMakePublic` as an install prefix, but I recommend: `<stage_dir>/CMAKE_INSTALL_PREFIX/lib/helloworld/cmake/external` (here `<stage_dir>` is where we are staging the install of HelloWorld and `CMAKE_INSTALL_PREFIX` is where the user wants the `CMakeSuperBuild` project installed. This means CMakePublic installs the header to `<stage_dir>/CMAKE_INSTALL_PREFIX/lib/helloworld/cmake/external/include/cmakepublic/cmake_public.hpp` and the library to `<stage_dir>/CMAKE_INSTALL_PREFIX/lib/helloworld/cmake/external/lib/libcmake_public.so`. Simply telling HelloWorld it depends on the `CMakePublic` target should get the paths right though.

Disclaimer: I made the executive decision to install the entire dependency to `<stage_dir>/CMAKE_INSTALL_PREFIX/lib/helloworld/cmake/external/`. Down the road what we probably want to do is incorporate a package manger under the hood of CMaize; once we do that, we'll let the package manager worry about these details. For now as long as we systematically install the dependencies it shouldn't be too hard to redirect them when the time for the package manager comes.

Commit:
Fixed include paths
- Root level CMakelIst looks at TheodoreDavis/CMakePublic for updated include paths
- Moved install location of depedencies to .../cmake/external/
- Added hello_world/include/helloworld path to mimic final install structure
-- Needed to avoid build vs install overhead in .cpp file
- Refactored hello-world.cpp include paths
- Refactored test.cpp include paths

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -2,37 +2,49 @@ cmake_minimum_required(VERSION 3.5.1)
-
-# Add cmake config location to paths to search
-list(APPEND CMAKE_PREFIX_PATH ${stage_directory})
+set(EXTERNAL_DIR "helloworld/cmake/external")
```
```suggestion
set(EXTERNAL_DIR "helloworld/external")
```
if we're going to actually put the external includes in `includes` and libraries in `libs` I think I like it better without the `cmake` prefix. When I recommended the `cmake` prefix I thought we were going to have to put everything in one directory. We can give this a shot, but I'm not too sure how widespread the use of `CMAKE_INSTALL_X` is with the exception of `X` equals `PREFIX`.

ryanmrichard committed:
Update CMakeLists.txt
